src/slopes.jl

```diff
@@ -0,0 +1,198 @@
+# Reference : Dietmar Ratz - An Optimized Interval Slope Arithmetic and its Application
+using IntervalArithmetic, ForwardDiff
+import Base: +, -, *, /, ^, sqrt, exp, log, sin, cos, tan, asin, acos, atan
+import IntervalArithmetic: mid, interval
+
+function slope{T}(f::Function, x::Interval{T}, c::T)
+    try
+        f(slope_var(x, c)).s
+    catch y
+        if isa(y, MethodError)
+            ForwardDiff.derivative(f, x)
+        end
+    end
+end
+
+struct Slope{T}
+    x::Interval{T}
+    c::Interval{T}
+    s::Interval{T}
+end
+
+Slope{T}(c) where T = Slope{T}(c, c, 0)
+
+function slope_var{T}(v::T)
+    Slope{T}(v, v, 1)
+end
+
+function slope_var{T}(v::Interval{T}, c::T)
+    Slope{T}(v, c, 1)
+end
+
+function interval(u::Slope)
+    u.x
+end
+
+function mid(u::Slope)
+    u.c
+end
+
+function slope(u::Slope)
+    u.s
+end
+
+function +{T}(u::Slope{T}, v::Slope{T})
+    Slope{T}(u.x + v.x, u.c + v.c, u.s + v.s)
+end
+
+function -{T}(u::Slope{T}, v::Slope{T})
+    Slope{T}(u.x - v.x, u.c - v.c, u.s - v.s)
+end
+
+function *{T}(u::Slope{T}, v::Slope{T})
+    Slope{T}(u.x * v.x, u.c * v.c, u.s * v.c + u.x * v.s)
+end
+
+function /{T}(u::Slope{T}, v::Slope{T})
+    Slope{T}(u.x / v.x, u.c / v.c, (u.s - (u.c / v.c) * v.s) / v.x)
+end
+
+function +{T}(u::Union{Interval{T}, Number}, v::Slope{T})
```

Comment:
Do these functions really need to be parameterised? And do they need type annotations for the first parameter?

Reply:
The functions do need to be parameterized since I'm making constructor calls in the functions that need the `convert` method to be called on them, and that would be done only if the parameterized constructor is called.
(Reference - https://docs.julialang.org/en/stable/manual/types/#Parametric-Composite-Types-1)
```
julia> Slope(1..2, 1..2, 1.5)
ERROR: MethodError: no method matching Slope(::IntervalArithmetic.Interval{Float64}, ::IntervalArithmetic.Interval{Float64}, ::Float64)
Closest candidates are:
  Slope(::IntervalArithmetic.Interval{T}, ::IntervalArithmetic.Interval{T}, ::IntervalArithmetic.Interval{T}) where T at none:2
Stacktrace:
 [1] macro expansion at /home/eeshan/.julia/v0.6/Atom/src/repl.jl:118 [inlined]
 [2] anonymous at ./<missing>:?

julia> Slope{Float64}(1..2, 1..2, 1.5)
Slope{Float64}([1, 2], [1, 2], [1.5, 1.5])
```

Reply:
You can just add a constructor Slope(a, b, c) that does the conversion.

Reply:
And for the type annotation, the behaviour of a function, say `+`, is different when adding two `Slope` types as compared to adding a `Slope` to an `Interval` or a  `Float`.
That's why I thought type annotations was necessary. Or should I attach annotations only for the function that adds two `Slope`s?